File: api-net/Services/OtpService.cs
using Liftonom.Api.Data;
using Liftonom.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace Liftonom.Api.Services;

public class OtpService(AppDbContext db, ISmsSender sms, IConfiguration config)
{
    private int Length => config.GetValue("Otp:Length", 6);
    private int TtlMinutes => config.GetValue("Otp:TtlMinutes", 3);
    private int MaxAttempts => config.GetValue("Otp:MaxAttempts", 3);

    public async Task<string> GenerateAsync(string phone, long? tenantId, string purpose = "login")
    {
        // Bekleyen eski kodları tüket
        await db.OtpCodes
            .Where(o => o.Phone == phone && o.Purpose == purpose && o.ConsumedAt == null)
            .ExecuteUpdateAsync(s => s.SetProperty(o => o.ConsumedAt, DateTime.UtcNow));

        var max = (int)Math.Pow(10, Length);
        var code = Random.Shared.Next(0, max).ToString().PadLeft(Length, '0');

        db.OtpCodes.Add(new OtpCode
        {
            TenantId = tenantId,
            Phone = phone,
            Code = code,
            Purpose = purpose,
            Attempts = 0,
            ExpiresAt = DateTime.UtcNow.AddMinutes(TtlMinutes),
        });
        await db.SaveChangesAsync();

        await sms.SendAsync(phone, $"Liftonom doğrulama kodunuz: {code}", "otp");
        return code;
    }

    public async Task<OtpCode> VerifyAsync(string phone, string code, string purpose = "login")
    {
        var otp = await db.OtpCodes
            .Where(o => o.Phone == phone && o.Purpose == purpose && o.ConsumedAt == null)
            .OrderByDescending(o => o.Id)
            .FirstOrDefaultAsync()
            ?? throw new ApiException(422, "Doğrulama kodu bulunamadı. Lütfen yeni kod isteyin.");

        if (otp.ExpiresAt < DateTime.UtcNow)
            throw new ApiException(422, "Doğrulama kodunun süresi doldu. Lütfen yeni kod isteyin.");

        if (otp.Attempts >= MaxAttempts)
            throw new ApiException(429, "Çok fazla hatalı deneme. Lütfen yeni kod isteyin.");

        if (otp.Code != code)
        {
            otp.Attempts++;
            await db.SaveChangesAsync();
            throw new ApiException(422, "Doğrulama kodu hatalı.");
        }

        otp.ConsumedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return otp;
    }
}

File: api/app/Services/OtpService.php
<?php
// app/Services/OtpService.php

namespace App\Services;

use App\Models\OtpCode;
use App\Services\Sms\SmsSender;
use Illuminate\Support\Carbon;

class OtpService
{
    public function __construct(private SmsSender $sms) {}

    /**
     * Yeni OTP üretir, kaydeder ve SMS ile gönderir.
     * Dönen kod yalnızca yerel/debug'da çağırana gösterilir.
     */
    public function generate(string $phone, ?int $tenantId = null, string $purpose = 'login'): string
    {
        $length = config('services.otp.length', 6);
        $ttl    = config('services.otp.ttl_minutes', 3);

        // Aynı amaç için bekleyen eski kodları tüket
        OtpCode::where('phone', $phone)
            ->where('purpose', $purpose)
            ->whereNull('consumed_at')
            ->update(['consumed_at' => now()]);

        $code = str_pad((string) random_int(0, (10 ** $length) - 1), $length, '0', STR_PAD_LEFT);

        OtpCode::create([
            'tenant_id'  => $tenantId,
            'phone'      => $phone,
            'code'       => $code,
            'purpose'    => $purpose,
            'attempts'   => 0,
            'expires_at' => Carbon::now()->addMinutes($ttl),
        ]);

        $this->sms->send($phone, "LiftOtonom doğrulama kodunuz: {$code}", 'otp');

        return $code;
    }

    /**
     * OTP doğrular. Başarılıysa tüketir ve true döner.
     * Hata durumunda kullanıcı dostu mesaj fırlatır.
     */
    public function verify(string $phone, string $code, string $purpose = 'login'): OtpCode
    {
        $otp = OtpCode::where('phone', $phone)
            ->where('purpose', $purpose)
            ->whereNull('consumed_at')
            ->latest('id')
            ->first();

        if (! $otp) {
            abort(422, 'Doğrulama kodu bulunamadı. Lütfen yeni kod isteyin.');
        }

        if ($otp->isExpired()) {
            abort(422, 'Doğrulama kodunun süresi doldu. Lütfen yeni kod isteyin.');
        }

        $max = config('services.otp.max_attempts', 3);
        if ($otp->attempts >= $max) {
            abort(429, 'Çok fazla hatalı deneme. Lütfen yeni kod isteyin.');
        }

        if (! hash_equals($otp->code, $code)) {
            $otp->increment('attempts');
            abort(422, 'Doğrulama kodu hatalı.');
        }

        $otp->update(['consumed_at' => now()]);

        return $otp;
    }
}
